### PR TITLE
wind.go: check for nil file while locking/unlocking.

### PR DESCRIPTION
**Pull request by knusbaum**
It seems that `(*Window).body.file` can be `nil` when a writing to an error window. I get the following panic when `Del`eting an error window while a program is writing to it:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0xb8 pc=0x1401cc2]

goroutine 50 [running]:
main.(*File).AllText(0x0, 0xc000095d18)
	/go/src/github.com/rjkroege/edwood/file.go:309 +0x22
main.(*Window).Lock(0xc000325800, 0x58)
	/go/src/github.com/rjkroege/edwood/wind.go:323 +0x8b
main.errorwin(0xc0000e4100, 0x58, 0x0)
	/go/src/github.com/rjkroege/edwood/util.go:154 +0x84
main.xfidwrite(0xc0000d8160)
	/go/src/github.com/rjkroege/edwood/xfid.go:438 +0x399
main.xfidctl(0xc0000d8160, 0x15cffe0, 0xc0001880e0)
	/go/src/github.com/rjkroege/edwood/xfid.go:39 +0x9c
created by main.xfidallocthread
	/go/src/github.com/rjkroege/edwood/acme.go:625 +0x221
```

It seems to be dependent on how fast the program is writing.
If I run the program below with Button 2 in a directory window (so output goes to `+Errors`), and then close the `+Errors` window, The panic above happens.

```go
package main

import (
	"fmt"
	"time"
)

func main() {
	var i int
	for {
		fmt.Printf("%d, ", i)
		i += 1
		time.Sleep(200 * time.Millisecond)
	}
}

```

Making sure that`w.body.file` is not `nil` prevents the panic.

### CODE REVIEW

**Comment by knusbaum** [comment]
May be related to #57
This fix may not be the best solution.

File: wind.go [hunk 1]
```diff
@@ -336,6 +339,9 @@ func (w *Window) unlock1() {
 
 // Unlock releases the lock on each clone of w
 func (w *Window) Unlock() {
+	if w.body.file == nil {
+		return
```
Comment (fhs) [inline wind.go:343]: Shouldn't we call `w.unlock1()` here before returning? We need undo these lines from the `Lock` method:
```go
w.lk.Lock()
w.ref.Inc()
w.owner = owner
```

Comment (knusbaum) [reply wind.go:343]: Yes, you're right.

**Comment by rjkroege** [comment]
The oops being my oops when I wrote that.

**Comment by rjkroege** [comment]
The test failure looks significant. Maybe @knusbaum you could have a look. Maybe I'm misreading the travis output but:

```
 go test -v -race -coverprofile=coverage.txt -covermode=atomic ./...
```

appears to haves timed out. That suggests a deadlock to me. Your PR has changed the order of unlock operations.

**Comment by knusbaum** [comment]
@rjkroege Hmm, I don't see how it reorders unlocking (I may be missing something)

I am suspicious that it may be possible for `w.body.file` to become `nil` between `Lock` and `Unlock` operations, which I think could cause problems. I haven't confirmed or even investigated that possibility though.

I'll look into this when I get time.

**Comment by rjkroege** [comment]
There's definitely bugs in this part of the code. Hopefully we'll get around to making it less fragile eventually.

**Comment by rjkroege** [comment]
w.r.t. 1191b06b3e56c539a6041ca1f0185ff39ff13a54

A few questions about the patch:

* when I run your patch, it times out. Is that the expected outcome? Did your merge / rebase
not work out? Also: please assume rebase workflow (i.e. can you make your pr without an
embedded merge op so that I can apply it via rebase?)
* edwood/draw/interface.go:53 how does `d.drawDisplay` ever end up `nil`?
* edwood/go.mod:23: what does `github.com/knusbaum/go v0.0.1` do?

I attempted to replicate the issue here as follows:

* middle-click on `yes`
* close the `+Errors` window

I get this crash log:
```
github.com/rjkroege/edwood/file.(*ObservableEditableBuffer).AllObservers(0x0, 0x14000345d10)
	/Users/rjkroege/tools/edwood/file/observable_editable_buffer.go:106 +0x20
main.(*Window).Lock(0x14000150400, 0x58)
	/Users/rjkroege/tools/edwood/wind.go:346 +0xd4
main.errorwin(0x140000c8100, 0x58)
	/Users/rjkroege/tools/edwood/util.go:95 +0x8c
main.xfidwrite(0x1400028e000)
	/Users/rjkroege/tools/edwood/xfid.go:436 +0x198
main.xfidctl(0x1400028e000, {0x102b58578, 0x140001b2118})
	/Users/rjkroege/tools/edwood/xfid.go:40 +0x8c
created by main.xfidallocthread
	/Users/rjkroege/tools/edwood/acme.go:535 +0x16c
testedwood.sh : exit status 2
```

`Window` locks appear to be correct. I think the problem
is in `errorwin`: we lookup a window here `edwood/util.go:66`.
It's happy to find a window from `g.row` while `g.row` is locked.

I think #420 fixes #57. Your observation that `f` was nil here was
very helpful.

**Comment by knusbaum** [comment]
Hi @rjkroege . 

I'm sorry, I forgot I had this PR open based on my fork's master branch. 

I've been doing some experimenting, which is what you're seeing here. It *is* broken right now, haha. Sorry to waste your time. I'll close this.

If I ever revisit the original issue I'll open a new PR.

**Pull request closed without merging**